File: languages/csharp_unity_json_properties/Ability.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Utils;

[System.Serializable]
public class Ability : Element
{
    [JsonProperty("activation"), TextAttribute("")]
    public string activation;
    [JsonProperty("duration"), Integer(0)]
    public int duration;
    [JsonProperty("potency"), Integer(100)]
    public int potency;
    [JsonProperty("range"), Integer(0)]
    public int range;
    [JsonProperty("effects"), ReferenceAttribute(typeof(Phenomenon), true)]
    public string effects;
    [JsonProperty("challenges"), TextAttribute("")]
    public string challenges;
    [JsonProperty("talents"), ReferenceAttribute(typeof(Trait), true)]
    public string talents;
    [JsonProperty("requisites"), ReferenceAttribute(typeof(Construct), true)]
    public string requisites;
    [JsonProperty("prevalence"), TextAttribute("")]
    public string prevalence;
    [JsonProperty("tradition"), ReferenceAttribute(typeof(Construct))]
    public string tradition;
    [JsonProperty("source"), ReferenceAttribute(typeof(Phenomenon))]
    public string source;
    [JsonProperty("locus"), ReferenceAttribute(typeof(Location))]
    public string locus;
    [JsonProperty("instruments"), ReferenceAttribute(typeof(Object), true)]
    public string instruments;
    [JsonProperty("systems"), ReferenceAttribute(typeof(Construct), true)]
    public string systems;
}

File: languages/django_models/ability.py
from .abstract_element_model import AbstractElementModel
from django.core.validators import MaxValueValidator
from django.db import models

class Ability(AbstractElementModel):

    # Mechanics
    activation = models.TextField(blank=True, null=True)
    duration = models.PositiveIntegerField(blank=True, null=True)
    potency = models.PositiveIntegerField(validators=[MaxValueValidator(100)], blank=True, null=True)
    range = models.PositiveIntegerField(blank=True, null=True)
    effects = models.ManyToManyField("Phenomenon", blank=True, related_name="ability_effects")
    challenges = models.TextField(blank=True, null=True)
    talents = models.ManyToManyField("Trait", blank=True, related_name="ability_talents")
    requisites = models.ManyToManyField("Construct", blank=True, related_name="ability_requisites")

    # World
    prevalence = models.TextField(blank=True, null=True)
    tradition = models.ForeignKey("Construct", on_delete=models.SET_NULL, blank=True, null=True, related_name="ability_tradition")
    source = models.ForeignKey("Phenomenon", on_delete=models.SET_NULL, blank=True, null=True, related_name="ability_source")
    locus = models.ForeignKey("Location", on_delete=models.SET_NULL, blank=True, null=True, related_name="ability_locus")
    instruments = models.ManyToManyField("Object", blank=True, related_name="ability_instruments")
    systems = models.ManyToManyField("Construct", blank=True, related_name="ability_systems")

    def __str__(self):
        return self.name
